src/circuits/noise.jl

```diff
@@ -0,0 +1,158 @@
+# Noise model gate definitions
+function gate(::GateName"1qubit-pauli_error"; pX::Number = 0.0, pY::Number = 0.0, pZ::Number = 0.0)
+  kraus = zeros(Complex{Float64},2,2,4)
+  kraus[:,:,1] = √(1-pX-pY-pZ) * gate("Id")
+  kraus[:,:,2] = √pX * gate("X")
+  kraus[:,:,3] = √pY * gate("Y")
+  kraus[:,:,4] = √pZ * gate("Z")
+  return kraus 
+end
+
+function gate(::GateName"2qubit-pauli_error"; probs = prepend!(zeros(15),1))
+  kraus = zeros(Complex{Float64},4,4,16)
+  basis = vec(reverse.(Iterators.product(fill(["Id","X","Y","Z"],2)...)|>collect))
+  for (k,ops) in enumerate(basis)
+    kraus[:,:,k] = √probs[k] * kron(gate(ops[1]), gate(ops[2]))
+  end
+  return kraus 
+end
+
+function gate(::GateName"pauli_error", N::Int = 1; kwargs...) 
+  N == 1 && return gate("1qubit-pauli_error"; kwargs...)
+  N == 2 && return gate("2qubit-pauli_error"; kwargs...)
+end
+
+
+gate(::GateName"pauli_channel"; kwargs...) = gate("pauli_error"; kwargs...)
+
+
+gate(::GateName"bit_flip"; p::Number) = 
+  gate("pauli_error"; pX = p)[:,:,1:2]
+  
+gate(::GateName"phase_flip"; p::Number) = 
+  gate("pauli_error"; pZ = p)[:,:,[1,4]]
+
+gate(::GateName"bit_phase_flip"; p::Number) = 
+  gate("pauli_error"; pY = p)[:,:,[1,3]]
+
+gate(::GateName"phase_bit_flip"; kwargs...) = 
+  gate("bit_phase_flip"; kwargs...)
+
+function gate(::GateName"AD"; γ::Number)
+  kraus = zeros(2,2,2)
+  kraus[:,:,1] = [1 0
+                  0 sqrt(1-γ)]
+  kraus[:,:,2] = [0 sqrt(γ)
+                  0 0]
+  return kraus 
+end
+
+# To accept the gate name "amplitude_damping"
+gate(::GateName"amplitude_damping"; kwargs...) = gate("AD"; kwargs...)
+
+function gate(::GateName"PD"; γ::Number)
+  kraus = zeros(2,2,2)
+  kraus[:,:,1] = [1 0
+                  0 sqrt(1-γ)]
+  kraus[:,:,2] = [0 0
+                  0 sqrt(γ)]
+  return kraus 
+end
+
+# To accept the gate name "phase_damping"
+gate(::GateName"phase_damping"; kwargs...) = gate("PD"; kwargs...)
+
+# To accept the gate name "dephasing"
+gate(::GateName"dephasing"; kwargs...) = gate("PD"; kwargs...)
+
+
+gate(::GateName"1qDEP"; p::Number) = 
+  gate("pauli_error"; pX = p/3.0, pY = p/3.0, pZ = p/3.0)
+
+gate(::GateName"2qDEP"; p::Number) = 
+  gate("pauli_error",2; probs = prepend!((p/15) * ones(15), 1-p))
+
+function gate(::GateName"DEP", N::Int = 1; kwargs...)
+  N == 1 && return gate("1qDEP"; kwargs...)
+  N == 2 && return gate("2qDEP"; kwargs...)
+end
+
+# To accept the gate name "depolarizing"
+gate(::GateName"depolarizing", N::Int = 1; kwargs...) = 
+  gate("DEP", N; kwargs...)
+
+
+function applynoise(circuit::Vector, noise::Tuple; kwargs...) 
+  applynoise(circuit, (noise1Q = noise, noise2Q = noise); kwargs...)
+end 
+
+function applynoise(circuit::Vector, noise::NamedTuple; idle = false, productnoise = false) 
```

Comment:
So is the point of this function to take a circuit and add some specified noise to it? I guess this is meant as an internal function? Otherwise we should come up with a better name. And what are `idle` and `productnoise`?

Reply:
To summarize our discussion: for now this could be an internal function until we work out the interface more (and the functionality could be accessed through `runcircuit`). Some ideas for changes:

1. Change `applynoise` to `insertnoise!` (and use `insert!` to actually modify the circuit in-plance, though it is also good to have a corresponding out-of-place version `insertnoise(circuit, [...]) = insertnoise!(copy(circuit), [...])`.
2. Rename `idle` to something more descriptive, like `idle_noise` or `spectator_noise`.
3. I think the plan was to remove `productnoise`?

Reply:
yes correct, I did not work on this just yet. Will have this ready for Monday

Reply:
No problem, just writing it down as a reminder before I forgot what we had discussed.